### PR TITLE
[ML] Apply source query on data frame analytics memory estimation

### LINKED ISSUE #49454: [ML] Data Frame Analytics memory estimation ignores Query
The Memory estimation process uses the extractor against the source index before the data frame analytics process runs. But, in doing this, it ignores the the data frame analytics source query. 

Extractor simply does a match_all: https://github.com/elastic/elasticsearch/blob/1dd816f030e256f497c1726695d1209a481b9d8e/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/extractor/DataFrameDataExtractorFactory.java#L42..L58

And the estimation process uses the extractor to determine row and column counts: https://github.com/elastic/elasticsearch/blob/12528b351a99941ffc1d44dd2c0265768ef79635/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/MemoryUsageEstimationProcessManager.java#L59..L60

The memory estimator needs to account for a defined query when estimating against the source index.

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
Closes #49454

### CODE REVIEW

File: x-pack/plugin/ml/qa/native-multi-node-tests/src/test/java/org/elasticsearch/xpack/ml/integration/ExplainDataFrameAnalyticsIT.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        }
+        BulkResponse bulkResponse = bulkRequestBuilder.get();
+        if (bulkResponse.hasFailures()) {
+            fail("Failed to index data: " + bulkResponse.buildFailureMessage());
+        }
+
+        String id = "test_source_query_is_applied";
+
+        DataFrameAnalyticsConfig.Builder configBuilder = new DataFrameAnalyticsConfig.Builder();
+        configBuilder.setId(id);
```
Comment (przemekwitek) [inline x-pack/plugin/ml/qa/native-multi-node-tests/src/test/java/org/elasticsearch/xpack/ml/integration/ExplainDataFrameAnalyticsIT.java]: `Builder` methods like `setId` return `this` so you can chain the calls.
This makes the builder configuration code cleaner IMO.

File: x-pack/plugin/ml/qa/native-multi-node-tests/src/test/java/org/elasticsearch/xpack/ml/integration/ExplainDataFrameAnalyticsIT.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        configBuilder.setSource(new DataFrameAnalyticsSource(new String[] { sourceIndex },
+            QueryProvider.fromParsedQuery(QueryBuilders.termQuery("categorical", "only-one"))));
+        configBuilder.setAnalysis(new Classification("categorical"));
+        DataFrameAnalyticsConfig config = configBuilder.buildForExplain();
+
+        ExplainDataFrameAnalyticsAction.Response explainResponse = explainDataFrame(config);
+
+        assertThat(explainResponse.getMemoryEstimation().getExpectedMemoryWithoutDisk().getKb(), lessThanOrEqualTo(500L));
```
Comment (przemekwitek) [inline x-pack/plugin/ml/qa/native-multi-node-tests/src/test/java/org/elasticsearch/xpack/ml/integration/ExplainDataFrameAnalyticsIT.java]: I'm wondering if 
```
assertThat(explainResponse.getMemoryEstimation().getExpectedMemoryWithoutDisk(), lessThanOrEqualTo(new ByteSizeValue(500, KB)));
```
would work as well.

**Review by przemekwitek — APPROVED** [review]
LGTM

Two minor comments inline.

File: x-pack/plugin/ml/qa/native-multi-node-tests/src/test/java/org/elasticsearch/xpack/ml/integration/ExplainDataFrameAnalyticsIT.java (hunk 2 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/qa/native-multi-node-tests/src/test/java/org/elasticsearch/xpack/ml/integration/ExplainDataFrameAnalyticsIT.java]: It does but it makes the line longer and I don't think it significantly improves readability.

**Comment by jpountz** [comment]
@dimitris-athanasiou FYI I'm removing the `v7.5.1` label as this doesn't seem to have been backported to the `7.5` branch.

**Comment by droberts195** [comment]
Looks like a logically equivalent fix is in 7.5.1, which is #49527.  Both PRs have the same title and #49527 is in the release notes.  Sorry for the extra work though.

**Pull request merged**